### PR TITLE
Feature/handle older tags

### PR DESCRIPTION
**Pull request by tuj**
* If CRC check fails, accept the tag if it does not follow standards for owner institution.

### CODE REVIEW

File: src/middleware/AbstractTagReader.java [hunk 1]
```diff
@@ -161,14 +161,14 @@ private String reverseData(String data) {
 	 * 
 	 * @see https://en.wikipedia.org/wiki/UTF-8
 	 */
-	private String utf8decode(String s) throws UnsupportedEncodingException {
+	private String utf8decode(String s, boolean breakAtNull) throws UnsupportedEncodingException {
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java]: Function parameter description needs some love.

File: src/middleware/AbstractTagReader.java [hunk 2]
```diff
@@ -178,6 +178,17 @@ private String utf8decode(String s) throws UnsupportedEncodingException {
...
 	}
 
+	/**
+	 * Decode a utf-8 hex encoded string and break at first null character.
+	 *
+	 * @param s
+	 * @return
+	 * @throws UnsupportedEncodingException
+	 */
+	private String utf8decode(String s) throws UnsupportedEncodingException {
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java]: Function parameter description needs some love.

File: src/middleware/AbstractTagReader.java [hunk 3]
```diff
@@ -223,6 +234,16 @@ private boolean crc(String crc, String data) {
 		return crc.equals(calculatedCrc);
 	}
 
+	/**
+	 * Check if owner institution matches ISIL.
+	 *
+	 * @param data
+	 * @return
+	 */
+	private boolean checkOwnerInstitution(String ownerInstitution) {
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java:257]: Function parameter description needs some love.

File: src/middleware/AbstractTagReader.java [hunk 4]
```diff
@@ -249,16 +272,17 @@ private void processTags(HashMap<String, BibTag> tags) {
 			// Make sure the tag has the correct order.
 			// @TODO: Can we get this info from somewhere else?
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java:301]: Is the todo valid still?

File: src/middleware/AbstractTagReader.java [hunk 5]
```diff
@@ -249,16 +272,17 @@ private void processTags(HashMap<String, BibTag> tags) {
...
 			if (data.substring(6, 8).equals("11")) {
+				logger.info("Ensuring tag order, reversing data in: " + data);
+
 				data = reverseData(data);
 			}
 
-			// We only accept tags that start with 11
+			// We only accept tags that start with 11 and has a data length of 64.
 			if (!data.substring(0, 2).equals("11") || data.length() != 64) {
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java:310]: No directly related by why "11" and whey "64" ?

File: src/middleware/AbstractTagReader.java [hunk 6]
```diff
@@ -267,16 +291,32 @@ private void processTags(HashMap<String, BibTag> tags) {
 				// Then we do not recognize the tag.
 				logger.warning("Could not validate tag: " + data);
 
-				iterator.remove();
-
-				continue;
+				// This is a hack to handle tags that do not follow standards and do not have valid CRC's.
+				// If it follows standards and does not have a valid CRC, ignore the tag.
+				try {
+					ownerInstitution = utf8decode(data.substring(42, 64), false);
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java:326]: Magic substring start and end values?

File: src/middleware/AbstractTagReader.java [hunk 7]
```diff
@@ -267,16 +291,32 @@ private void processTags(HashMap<String, BibTag> tags) {
...
+					if (checkOwnerInstitution(ownerInstitution)) {
+						iterator.remove();
```
Comment (cableman) [inline src/middleware/AbstractTagReader.java:329]: Nice with comment about remove... when owner valid?

**Review by cableman — CHANGES_REQUESTED** [review]
Minor stuff

File: src/middleware/AbstractTagReader.java (hunk 7 above)
Comment (tuj) [reply src/middleware/AbstractTagReader.java:329]: Comment added

File: src/middleware/AbstractTagReader.java (hunk 6 above)
Comment (tuj) [reply src/middleware/AbstractTagReader.java:326]: Description of tag layout added to top of method.

File: src/middleware/AbstractTagReader.java (hunk 5 above)
Comment (tuj) [reply src/middleware/AbstractTagReader.java:310]: Expanded comment

File: src/middleware/AbstractTagReader.java (hunk 4 above)
Comment (tuj) [reply src/middleware/AbstractTagReader.java:301]: Yes

File: src/middleware/AbstractTagReader.java (hunk 3 above)
Comment (tuj) [reply src/middleware/AbstractTagReader.java:257]: Done

File: src/middleware/AbstractTagReader.java (hunk 2 above)
Comment (tuj) [reply src/middleware/AbstractTagReader.java]: Done

File: src/middleware/AbstractTagReader.java (hunk 1 above)
Comment (tuj) [reply src/middleware/AbstractTagReader.java]: Done

**Comment by tuj** [comment]
On accept, the following PR should be merged: https://github.com/bibboxen/bibbox/pull/25

**Pull request merged**